---
title: "R Notebook"
output: html_notebook
---


```{r load packages}
library(dplyr)
library(tidyr)
library(zoo)
library(ggplot2)
library(ggpubr)
library(ggfortify)
library(SNFtool)
library(data.table)
library(RColorBrewer)
library(plyr)
library(vcd)
library(UpSetR)
library(MASS)
library(alluvial)
library(ggalluvial)
```

```{r load snf data}
# File path on scc: /external/rprshnas01/kcni/mwang/rosmap/
rosmap_path<-'C:/Users/yangmu/Desktop/rosmap/'
Data_rnaseq<-read.csv(file.path(rosmap_path,'counts_adj_ready.csv'),row.names = 1, header = TRUE)

resid_path<-'C:/Users/yangmu/Desktop/rosmap/SNF sample analysis'
Data_chipseq<-read.csv(file.path(resid_path,'histone_resids.csv'),row.names = 1, header = TRUE)
Data_methylation<-read.csv(file.path(resid_path,'dna_resids.csv'),row.names = 1, header = TRUE)
Data_tmt <- read.csv(file.path(resid_path,'tmt_resids.csv'),row.names = 1, header = TRUE)
Data_metabo <- read.csv(file.path(resid_path,'metabo_resids.csv'),row.names = 1, header = TRUE)


#####MODIFY DATAFRAMES################
rnaseq_col <- rownames(Data_rnaseq)
rnaseq_row <- colnames(Data_rnaseq)
Data_rnaseq<-t(Data_rnaseq)
rownames(Data_rnaseq)<- rnaseq_row
colnames(Data_rnaseq)<- rnaseq_col


#############Data frames are in the format of patient x features going forward##################
```



```{r set parameters}
#############################
# try different settings; the medulloblastoma study use k and t up to 50
K=40 # number of neibors
alpha = 0.5 #hyperparameter
T=50 #number of iteration
#############################
```


```{r define functions}
# Function of find mode of a vector
# if mode does not exist, return the first element
Mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

# Function of defining optimal cluster number
est_clust <- function(x){
  test<-SNFtool::estimateNumberOfClustersGivenGraph(x,NUMC = 2:8)
  #print(test)
  # If optimal cluster numbers are same under two default methods
  if (test[[1]]==test[[3]]){
    test[[1]]
  } 
  # If not, we want to use the mode of solutions, but need to avoid 
  # the case when mode is created by second best solutions
  else if (test[[2]]==test[[4]]) {
    min(test[[1]],test[[3]])
  } 
  # calculate mode if exist
  else if (length(unique(test)<4)){
    as.integer(Mode(test))
  }
  # If mode does not exist, use the minimum number of optimal solutions
  else {min(test[[1]],test[[3]])}
}
```


```{r data normalization/distance matrix}
# Calculate stardardized normalized matrices and distance matrices

# Remove empty values due to missing pmi, age_death, study or sex
Data_methylation <- Data_methylation[complete.cases(Data_methylation), ]
Data_chipseq <- Data_chipseq[complete.cases(Data_chipseq), ]
Data_tmt <- Data_tmt[complete.cases(Data_tmt), ]


## RNA-seq
rnaseq_s<-standardNormalization(Data_rnaseq)
rnaseq_d<-(dist2(as.matrix(rnaseq_s),as.matrix(rnaseq_s)))^(1/2)

## DNA-methylation
methy_s<-standardNormalization(Data_methylation)
methy_d<-(dist2(as.matrix(methy_s),as.matrix(methy_s)))^(1/2)

## ChIP-seq
chipseq_s<-standardNormalization(Data_chipseq)
chipseq_d<-(dist2(as.matrix(chipseq_s),as.matrix(chipseq_s)))^(1/2)

## Proteomics-tmt (Random forest imputed)
prot_s<-standardNormalization(Data_tmt)
prot_d<-(dist2(as.matrix(prot_s),as.matrix(prot_s)))^(1/2)

## Metaboloics (Random forest imputed)
metabo_s<-standardNormalization(Data_metabo)
metabo_d<-(dist2(as.matrix(metabo_s),as.matrix(metabo_s)))^(1/2)

```

```{r full sample}
# get the ids for common samples
chipseq_row <- rownames(Data_chipseq)
methy_row <- rownames(Data_methylation)
meta_row <- rownames(Data_metabo)
tmt_row <- rownames(Data_tmt)
full_sample <-Reduce(intersect,list(chipseq_row,methy_row,rnaseq_row,tmt_row,meta_row))
# or you can match it with the cluster result derived previously from RDHPM_limitby_RDHPM

```


```{r snf fusing}
snf_sample <- full_sample
snf_list <- list()
d_list <- list(rnaseq_d,methy_d,chipseq_d,metabo_d,prot_d)
for (i in 1:5){
  snf_list[[i]]<-d_list[[i]][(row.names(d_list[[i]]) %in% snf_sample),(colnames(d_list[[i]]) %in% snf_sample)]
  snf_list[[i]]<-affinityMatrix(snf_list[[i]],K,alpha)
    # reorder patients id
  snf_list[[i]] <- snf_list[[i]][snf_sample,snf_sample]
}
snf = SNF(snf_list,K,T)
```



```{r get it from clvalid}
# Clvalid function for calculating APN, AD, ADM, FOM
stability <- function(mat, Dist=NULL, del, cluster, clusterDel, method="euclidean") {

  any.na <- any(is.na(mat))
  obsNum <- 1:nrow(mat)
  nc1 <- length(table(cluster))
  nc2 <- length(table(clusterDel))
  stabmeas <- numeric(4)
  names(stabmeas) <- c("APN","AD","ADM","FOM")

  ## measure APN
  ## calculate a ncxnc matrix of proportion of non-overlaps in the two collection of nc clusters
  overlap <- xtabs(~cluster + clusterDel)
  ## measure AD
  ## calculate a ncxnc matrix of average-distance in the two collection of nc clusters
  dij <- matrix(rep(NA,nc1*nc2),nc1,nc2)

  if (is.null(Dist)) matDist <- as.matrix(dist(mat, method=method))
  if ("dist" %in% class(Dist)) matDist <- as.matrix(Dist)
  if ("matrix" %in% class(Dist)) matDist <- Dist

  ## measure ADM
  ## calculate a ncxnc matrix of distance-average in the two collection of nc clusters
  dij2 <- matrix(rep(NA,nc1*nc2),nc1,nc2)
  ii <- 1
  for (i in sort(unique(cluster))) {
    jj <- 1
    xbari <- apply(mat[cluster==i,,drop=FALSE],2, function(x) mean(x, na.rm=TRUE))
    for (j in sort(unique(clusterDel))) {
      ## measure AD
      clusi <- obsNum[cluster==i]
      clusdelj <- obsNum[clusterDel==j]
      cl <- length(clusi)*length(clusdelj)
      if (cl>0) dij[ii,jj] <- mean(matDist[clusi,clusdelj], na.rm=TRUE)
      ##      if (cl>0) dij[ii,jj] <- mean(as.matrix(Dist)[clusi,clusdelj])
      ## measure ADM
      xbarj <- apply(mat[clusterDel==j,,drop=FALSE],2, function(x) mean(x, na.rm=TRUE))
      diff <- xbari-xbarj
      if(length(diff)>0) {
        if(any.na) {
          diff <- diff[!is.na(diff)]
          dij2[ii,jj] <- sqrt(mean(diff^2))
        } else {
          dij2[ii,jj] <- sqrt(sum(diff^2))
        }
      } else {
        dij2[ii,jj] <- 0
      }
      jj <- jj+1
    }
    ii <- ii+1
  }
  rs <- matrix(rowSums(overlap),nrow=nrow(overlap),ncol=ncol(overlap),byrow=FALSE)
  cs <- matrix(colSums(overlap),nrow=nrow(overlap),ncol=ncol(overlap),byrow=TRUE)
  stabmeas["APN"] <- 1-sum(overlap^2/rs)/sum(overlap)
  stabmeas["AD"] <- sum(overlap*dij)/nrow(mat)
  stabmeas["ADM"] <- sum(overlap*dij2)/nrow(mat)
  return(stabmeas)
}

```

```{r clvalid simulation}
# modified on clvalid codes
## take snf affinity matrix and number of clusters as input
stabmeas_table <- function(snf, k) {

  # Create a data frame to store stats
  measNames <- c("APN","AD","ADM","FOM")
  n <- ncol(snf)
  measures <- matrix(0,nrow=length(measNames),ncol=n)
  rownames(measures) <- measNames
  
  # perform spectral clustering on full data
  mat <- snf
  cluster <- spectralClustering(mat,k)
  # store cluster result
  names(cluster) <- rownames(mat)
  
  # iteration on column removal
  for (del in 1:ncol(mat)) {
    # remove 1 individuals at a time from affinity matrix
    matDel <- mat[-del,-del]   
    
    # Perform specteral clustering and stroe result
    clusterDel <- spectralClustering(matDel,k)
    names(clusterDel) <- rownames(matDel)
    cluster_input <- cluster[-del]
    
    # call the stability function
    stabmeas <- stability(mat, Dist=NULL, del, cluster=cluster_input, clusterDel)
    
    # write data
    measures["APN",del] <- measures["APN",del] + stabmeas["APN"]
    measures["AD",del]  <- measures["AD",del]  + stabmeas["AD"]
    measures["ADM",del] <- measures["ADM",del] + stabmeas["ADM"]
    #measures["FOM",del] <- measures["FOM",del] + stabmeas["FOM"]
    
  }
  # return a table with output
  return(measures)
}
```

```{r stability on diff cluster}
# creat a table to store output
output <- matrix(ncol=7, nrow=4)
rownames(output) <- c("APN","AD","ADM","FOM")
colnames(output) <- c(2:8)

# perform stability analysis on each of these cluster numbers
for (i in 2:8){
  output[,i-1] <- rowMeans(stabmeas_table(snf,i))
  print(i)
}
```

```{r plot result}
plot(colnames(output),output["APN",],  type = "b", pch = 19, col = "black", xlab = "cluster number", ylab = "APN",main="Cluster Stability (APN)")
plot(colnames(output),output["AD",],  type = "b", pch = 19, col = "black", xlab = "cluster number", ylab = "AD",main="Cluster Stability (AD)")
plot(colnames(output),output["ADM",],  type = "b", pch = 19, col = "black", xlab = "cluster number", ylab = "ADM",main="Cluster Stability (ADM)")
```





# previous results/methods

```{r Clvalid internal validation}
library(clValid)
set.seed(123)
intern <- clValid(metabo_s, 2:8, clMethods=c("kmeans","pam"),
                  validation="internal")

## view results
summary(intern)
optimalScores(intern)
plot(intern)
```


```{r Clvalid stability}
library(clValid)
stab <- clValid(metabo_s, 2:8, clMethods=c("kmeans","pam"),
                validation="stability")
optimalScores(stab)
plot(stab)
```

```{r plot sil}
library(cluster)
cluster_ouput <- intern@clusterObjs$kmeans$`2`$cluster
dmat <- daisy(metabo_s)
test<-silhouette(cluster_ouput, dist =dmat)
plot(test,col=1:8)
```


```{r silhouette from cluster}
library(cluster)
alluv_path<-"C:/Users/yangmu/Desktop/rosmap/SNF sample analysis/alluv_df"
coln <- c('clustnum2','clustnum3','clustnum4','clustnum5')
full<-read.csv(file.path(alluv_path,"RDHPM_limitby_RDHPM.csv"),  header = TRUE)

three <- read.csv(file.path(alluv_path,"RDH_limitby_RDH.csv"),  header = TRUE)
#full[coln] <- lapply(full[coln], factor)
full <- full %>%
  dplyr::rename(clustering = clustnum2)

three <- three %>%
  dplyr::rename(clustering = clustnum5) 

cluster_ouput <- full$clustering
snf_sil <- snf[full$individualID,full$individualID]*2
test<-silhouette(cluster_ouput, dmatrix=1-snf_sil)
plot(test)
```


```{r kmeans result}
kmeans<-intern@clusterObjs$kmeans$`2`$cluster
kmeans_df <- stack(kmeans)
kmeans_df <- kmeans_df %>%
  dplyr::rename(kmeans = values,
                individualID = ind) %>%
  dplyr::mutate(kmeans = as.factor(kmeans))

snf_output <- full %>%
  dplyr::select(individualID,clustering)%>%
  dplyr::rename(spectral = clustering) %>%
  dplyr::mutate(spectral = as.factor(spectral))


check <- inner_join(snf_output, kmeans_df, by='individualID')

check %>%
    group_by(spectral,kmeans) %>%
    dplyr::mutate(freq=n()) %>%
    slice(1)%>%
    ungroup %>%
    ggplot(aes(y=freq,axis1=spectral,axis2=kmeans)) + 
    geom_alluvium(aes(fill = spectral))+
    geom_stratum(width = 1/12, fill = "white", color = "grey")+
    geom_label(stat = "stratum", aes(label = after_stat(stratum))) +
    scale_x_discrete(limits = c('spectral','kmeans'), expand = c(.05, .05)) + 
    labs(title = 'Cluster Validation - Fully Integrated Subtypes')
chisq.test(as.matrix(table(check$spectral,check$kmeans)))
```

```{r}
set.seed(123)
p_list <- c()
accuracy_list <- c()
j_list <- c()
for (i in 1:100) {
  snf_sample <- sample(full_sample,size = 88)
  snf_list <- list()
  for (i in 1:5){
    snf_list[[i]]<-d_list[[i]][(row.names(d_list[[i]]) %in% snf_sample),(colnames(d_list[[i]]) %in% snf_sample)]
    snf_list[[i]]<-affinityMatrix(snf_list[[i]],K,alpha)
    # reorder patients id
    snf_list[[i]] <- snf_list[[i]][snf_sample,snf_sample]
  }
  snf = SNF(snf_list,K,T)
  # estimate optimal cluster number
  opt_est <- 2
  print(opt_est)
  # spectral clustering and attached to ID's
  output<-spectralClustering(snf,opt_est)
  
  
  cat_var <- data.frame(individualID=snf_sample,output)
  cat_var$output <- as.factor(cat_var$output)
  
  test_table <- inner_join(full_clust,cat_var, by= 'individualID')
  
  p<-fisher.test(table(test_table$full_subtype,test_table$output))$p.value
  p_list<-append(p_list,p)
  
  c_matrix <- caret::confusionMatrix(data=test_table$full_subtype, reference = test_table$output)
  accuracy <- c_matrix$overall['Accuracy']
  accuracy_list<-append(accuracy_list,max(accuracy,1-accuracy))
  
  jaccard <- clusteval::jaccard(test_table$full_subtype,test_table$output)
  j_list <- append(j_list, jaccard)
}

```
  
