---
title: "R Notebook"
output: html_notebook
---


```{r load packages}
library(dplyr)
library(tidyr)
library(zoo)
library(ggplot2)
library(ggpubr)
library(ggfortify)
library(SNFtool)
library(data.table)
library(RColorBrewer)
library(plyr)
library(vcd)
library(UpSetR)
library(MASS)
library(alluvial)
library(ggalluvial)
```

```{r load snf data}
# File path on scc: /external/rprshnas01/kcni/mwang/rosmap/
rosmap_path<-'C:/Users/yangmu/Desktop/rosmap/'
Data_chipseq<-read.csv(file.path(rosmap_path,'Chipseq_ready.csv'),row.names = 1, header = TRUE)
Data_methylation<-read.csv(file.path(rosmap_path,'Methylation_ready.csv'),row.names = 1, header = TRUE)
Data_rnaseq<-read.csv(file.path(rosmap_path,'counts_adj_ready.csv'),row.names = 1, header = TRUE)
Data_tmt <- read.csv(file.path(rosmap_path,'tmt_ready.csv'),row.names = 1, header = TRUE)
Data_metabo <- read.csv(file.path(rosmap_path,'metabolon_ready.csv'),row.names = 1, header = TRUE)


#####MODIFY DATAFRAMES################
rnqseq_col <- rownames(Data_rnaseq)
rnaseq_row <- colnames(Data_rnaseq)
Data_rnaseq<-t(Data_rnaseq)
rownames(Data_rnaseq)<- rnaseq_row
colnames(Data_rnaseq)<- rnqseq_col

methy_col <- rownames(Data_methylation)
methy_row <- colnames(Data_methylation)
Data_methylation<-t(Data_methylation)
rownames(Data_methylation)<- methy_row
colnames(Data_methylation)<- methy_col

chipseq_col <- rownames(Data_chipseq)
chipseq_row <- colnames(Data_chipseq)
Data_chipseq<-t(Data_chipseq)
rownames(Data_chipseq)<- chipseq_row
colnames(Data_chipseq)<- chipseq_col

tmt_col <- rownames(Data_tmt)
tmt_row <- colnames(Data_tmt)
Data_tmt<-t(Data_tmt)
rownames(Data_tmt)<- tmt_row
colnames(Data_tmt)<- tmt_col

meta_col <- rownames(Data_metabo)
meta_row <- colnames(Data_metabo)
Data_metabo<-t(Data_metabo)
rownames(Data_metabo)<- meta_row
colnames(Data_metabo)<- meta_col

#############Data frames are in the format of patient x features going forward##################

#############Feature Reduction for DNA-methylation data#########################
# calculate variance for each feature (columns)
var_methylation <- apply(Data_methylation, 2, var)
# set a threshold of variance (look at feature numbers)
var_methylation <- var_methylation[var_methylation>0.002]
# feature selection
selected_features<-attr(var_methylation, "names")
Data_methylation <- Data_methylation[,(colnames(Data_methylation) %in% selected_features)]

```

```{r set parameters}
#############################
# try different settings; the medulloblastoma study use k and t up to 50
K=40 # number of neibors
alpha = 0.5 #hyperparameter
T=50 #number of iteration
#############################
```


```{r dimension reduction based on normalized matrices}
########## DON'T RUN THIS CHUNK ####################
#var_methylation <- apply(Data_methylation, 2, var)
#var_methylation <- var_methylation[var_methylation>0.001]
#selected_features<-attr(var_methylation, "names")
#Data_methylation_s <- Data_methylation[,(colnames(Data_methylation) %in% selected_features)]

# Reduced Distance matrix 
#methy_ns<-standardNormalization(Data_methylation_s)
#methy_ds<-(dist2(as.matrix(methy_ns),as.matrix(methy_ns)))^(1/2)
# Non Reduced
#methy_n<-standardNormalization(Data_methylation)
#methy_d<-(dist2(as.matrix(methy_n),as.matrix(methy_n)))^(1/2)


# affinity
#w_methy<-affinityMatrix(methy_d,K,alpha)
#w_methys<-affinityMatrix(methy_ds,K,alpha)

# check cluster effects
#estimateNumberOfClustersGivenGraph(w_methy,NUMC = 2:10)
#estimateNumberOfClustersGivenGraph(w_methys,NUMC = 2:10)
#cluster = spectralClustering(w_methy,3)
#s_cluster = spectralClustering(w_methys,3)
#df_for_alluv <- data.frame(cluster,s_cluster)
#alluv_ready <- df_for_alluv %>%
#  group_by(cluster,s_cluster) %>%
#  dplyr::mutate(freq=n()) %>%
#  slice(1)%>%
#  ungroup() 
#alluvial(alluv_ready[,1:2],freq=alluv_ready$freq,
#         alpha = 0.7)
########## DON'T RUN THIS CHUNK ####################
```


```{r define functions}
# Function of find mode of a vector
# if mode does not exist, return the first element
Mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

# Function of defining optimal cluster number
est_clust <- function(x){
  test<-estimateNumberOfClustersGivenGraph(x,NUMC = 2:8)
  # If optimal cluster numbers are same under two default methods
  if (test[[1]]==test[[3]]){
    test[[1]]
  } 
  # If not, we want to use the mode of solutions, but need to avoid 
  # the case when mode is created by second best solutions
  else if (test[[2]]==test[[4]]) {
    min(test[[1]],test[[3]])
  } 
  # calculate mode if exist
  else if (length(unique(test)<4)){
    as.integer(Mode(test))
  }
  # If mode does not exist, use the minimum number of optimal solutions
  else {min(test[[1]],test[[3]])}
}
```


```{r data normalization/distance matrix}
# Calculate stardardized normalized matrices and distance matrices

## RNA-seq
rnaseq_d<-standardNormalization(Data_rnaseq)
rnaseq_d<-(dist2(as.matrix(rnaseq_d),as.matrix(rnaseq_d)))^(1/2)

## DNA-methylation
methy_d<-standardNormalization(Data_methylation)
methy_d<-(dist2(as.matrix(methy_d),as.matrix(methy_d)))^(1/2)

## ChIP-seq
chipseq_d<-standardNormalization(Data_chipseq)
chipseq_d<-(dist2(as.matrix(chipseq_d),as.matrix(chipseq_d)))^(1/2)

## Proteomics-tmt (Random forest imputed)
prot_d<-standardNormalization(Data_tmt)
prot_d<-(dist2(as.matrix(prot_d),as.matrix(prot_d)))^(1/2)

## Metaboloics (Random forest imputed)
metabo_d<-standardNormalization(Data_metabo)
metabo_d<-(dist2(as.matrix(metabo_d),as.matrix(metabo_d)))^(1/2)

```



```{r load regression data}
# Load Data (which will be used in regression)
phenotype_basic <- read.csv(file.path(rosmap_path,'phenotype_basic.csv'), header = TRUE)
data_clinical <- read.csv(file.path(rosmap_path,'ROSMAP_clinical.csv'), header = TRUE)

# setting variables for regression model
regression <- phenotype_basic %>%
  dplyr::select(projid,msex,age_death,pmi,apoe_genotype,
                amyloid,tangles) %>%
  mutate(amyloid_sqrt = sqrt(amyloid),
         tangles_sqrt = sqrt(tangles)
         )
regression$msex <- as.factor(regression$msex)
regression$apoe4 <- as.factor(ifelse(regression$apoe_genotype %in% c(24,34,44),
                            1,0))

# create a 2-column-df for mapping 2 types of ID's
id_map <- data_clinical %>%
  dplyr::select(projid,individualID)
```


```{r creat parameter space}
# Data types: RNA-seq(R), DNA-methylation(D), ChIP-seq(H), Proteomics-tmt(P), Metabolomics(M)

# define a vector of data types, which will be used to locate matrices
d_types <- c('R','D','H','P','M')

# Define a parameter space
param_space <- c()
for (i in 1:length(d_types)){
  m <-combn(d_types,i)
  num <- dim(m)[2]
  for (j in 1:num){
    param <- paste(m[,j],collapse='')
    param_space<- append(param_space,param)
  }
}

# size of the parameter space, which will be looped over
space_length <-length(param_space)
```

```{r define an empty datafram to write}
# define an empty data frame 
param_df <- data.frame(limit_by=character(), # given datatypes that limits sample size
                 clust_num=integer(), # cluster number used (looped from 2 to 5)
                 opt_est = integer(), # optimal cluster number given limit_by and snf_type
                 snf_type=character(), # datatypes used in SNF, subset of limit_by
                 sample = integer(), # sample size
                 response = character(), # response variable
                 p = numeric(), # p-value from F-test
                 stringsAsFactors=FALSE) 
```


```{r actual code}
# create a list of distance matrices
d_list <- list(rnaseq_d,methy_d,chipseq_d,prot_d,metabo_d)
# define a parameter space for cluster number
cluster_space <- c(2,3,4,5)
# define a parameter for response variable
response <- c('amyloid','tangles')
# define a list of response variable + 'output'
reg_param <- c('amyloid_sqrt','tangles_sqrt','output')
  
  
  
# The first layer of loop will be performed over parameter space
for (i in 1:space_length){
  # sample size is limited by elements in parameter space (limit_by)
  l_types <- param_space[i]
  # separate joint datatypes ('RD' to 'R' 'D' for example)
  p_limit <- as.vector(strsplit(l_types,'')[[1]])
  # update this list  of id based on joint datatypes
  limit_list <- list()
  # a parameter that locate datatypes accordingly
  pos <- c(0,0,0,0,0)
  # extract common patients and add position variable
  for (i in 1:length(d_types)) {
    if (d_types[i] %in% p_limit) {
      # update with individual ID's
      limit_list[[i]] <- rownames(d_list[[i]])
      # update the position/dummy vector
      pos[i]=1
    }
  }
  # Now we do not need positions for ID's, reduce the length of the list
  limit_list <- limit_list[!sapply(limit_list,is.null)]
  # take intersection of the id
  limit_id <- Reduce(intersect,limit_list)
  # extract sample size
  size <- length(limit_id)
  
  # define the sub-datatypes, which will be used for SNF (snf_types)
  max_l <- length(p_limit)
  # define a empty list for sub-datatype parameter space
  limit_space <- c()
  for (i in 1:max_l){
    m <- combn(p_limit,i)
    num <- dim(m)[2]
    for (j in 1:num){
    param <- paste(m[,j],collapse='')
    limit_space<- append(limit_space,param)
    }
  }
  # define the length of sub-datatype parameter space, looped by snf
  limit_l <- length(limit_space)
  print(limit_l)
  
  for (i in 1:limit_l){
    # snf only tooks one joint datatype at a time
    snf_types <- limit_space[i]
    # seperate the joint datatype
    limit <- strsplit(snf_types,'')[[1]]
    # another position variable defined for snf
    pos_sub <- c(0,0,0,0,0)
    
    # update position variable for sub-datatypes
    # no need for id this time
    for (i in 1:length(d_types)) {
      if (d_types[i] %in% limit) {
        pos_sub[i]=1
      }
    }
    
    # define snf inputs (list of affinity matrices)
    snf_list <- list()
    
    # update snf inputs
    for (i in 1:5){
      if (pos_sub[i]==1){
        # only took col/rows for common patients
        snf_list[[i]]<-d_list[[i]][(row.names(d_list[[i]]) %in% limit_id),(colnames(d_list[[i]]) %in% limit_id)]
        # calculate affinity matrix
        snf_list[[i]]<-affinityMatrix(snf_list[[i]],K,alpha)
        # reorder patients id
        snf_list[[i]] <- snf_list[[i]][limit_id,limit_id]
        #print(dim(snf_list[[i]]))
      }
    }
    
    # Remove null component from empty sub-position
    snf_list <- snf_list[!sapply(snf_list,is.null)]
    
    # 2 cases:  datatype (no need for integration) vs multiple datatypes
    if (sum(pos_sub) == 1){
      for (i in 1:length(cluster_space)){
        clust_num = cluster_space[i]
        # estimate optimal cluster number
        opt_est <- est_clust(snf_list[[1]])
        # spectral clustering and attached to ID's
        output<-spectralClustering(snf_list[[1]],clust_num)
        cat_var <- data.frame(individualID=limit_id,output)
        
        # Merge SNF output to regression data
        cat_var <- left_join(cat_var,id_map,by='individualID')
        reg_part <- left_join(cat_var,regression,by ='projid')
        reg_part['output']<-lapply(reg_part['output'],factor)
        reg_part <- reg_part[,reg_param]
        
        for (i in 1:length(response)){
          # tryCatch to skip errors due to cluster size
          tryCatch({
            sum_m <-summary(aov(data = reg_part, reg_part[,i] ~ output))
          # derive p-value
          p_val <- sum_m[[1]]["Pr(>F)"][[1]][1]
          # update data frame
          new_row <- c(l_types,clust_num,opt_est,snf_types,size,response[i],p_val)
          param_df[nrow(param_df)+1,] <- new_row
          },error = function(e){})
        }
      }
    } 
    
    # multiple datatypes (SNF Required)
    else{
      for (i in 1:length(cluster_space)){
        clust_num = cluster_space[i]
        # snf integration
        snf = SNF(snf_list,K,T)
        # estimate optimal cluster number
        opt_est <- est_clust(snf)
        # spectral clustering and attached to ID's
        output<-spectralClustering(snf,clust_num)
        cat_var <- data.frame(individualID=limit_id,output)
        
        # Merge SNF output to regression data
        cat_var <- left_join(cat_var,id_map,by='individualID')
        reg_part <- left_join(cat_var,regression,by ='projid')
        reg_part['output']<-lapply(reg_part['output'],factor)
        reg_part <- reg_part[,reg_param]
        
        for (i in 1:length(response)){
          # tryCatch to skip errors due to cluster size
          tryCatch({
            sum_m <-summary(aov(data = reg_part, reg_part[,i] ~ output))
          # derive p-value
          p_val <- sum_m[[1]]["Pr(>F)"][[1]][1]
          # update data frame
          new_row <- c(l_types,clust_num,opt_est,snf_types,size,response[i],p_val)
          param_df[nrow(param_df)+1,] <- new_row
          },error = function(e){})
        }
      }
    }
  }
}
# safe tge param_df dataframe for visualization
write.csv(param_df,"C:/Users/yangmu/Desktop/rosmap/SNF sample analysis/param_df_0411.csv", row.names = TRUE)
```
